presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpGenericPattern.java

```diff
@@ -0,0 +1,406 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.common.primitives.Ints;
+import com.google.re2j.Options;
+import com.google.re2j.Options.EventsListener;
+import com.google.re2j.Pattern;
+import io.airlift.jcodings.specific.NonStrictUTF8Encoding;
+import io.airlift.joni.Option;
+import io.airlift.joni.Regex;
+import io.airlift.joni.Region;
+import io.airlift.joni.Syntax;
+import io.airlift.joni.exception.ValueException;
+import io.airlift.log.Logger;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+
+import java.nio.charset.StandardCharsets;
+
+import static com.facebook.presto.operator.scalar.RegexpGenericPattern.RegexLibrary.JONI;
+import static com.facebook.presto.operator.scalar.RegexpGenericPattern.RegexLibrary.RE2J;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class RegexpGenericPattern
+{
+    private static final String DOT_STAR = ".*";
+
+    private static final Logger log = Logger.get(RegexpGenericPattern.class);
+
+    public enum RegexLibrary
+    {
+        RE2J,
+        JONI
+    }
+
+    public final RegexLibrary regexLibrary;
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public Pattern re2jPattern;
+    public Pattern re2jPatternWithoutDotStartPrefixSuffix;
+    public Regex joniRegex;
+
+    public RegexpGenericPattern(RegexLibrary regexLibrary, int dfaStatesLimit, int dfaRetries)
+    {
+        this.regexLibrary = regexLibrary;
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+    }
+
+    public void createPatternObject(Slice pattern)
+            throws Exception
+    {
+        if (regexLibrary == RE2J) {
+            Options options = Options.builder()
+                    .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                    .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                    .setNumberOfDFARetries(dfaRetries)
+                    .setEventsListener(new RE2JEventsListener())
+                    .build();
+
+            String patternString = pattern.toStringUtf8();
+            re2jPattern = Pattern.compile(patternString, options);
+
+            // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+            String patternStringWithoutDotStartPrefixSuffix = patternString;
+            if (patternStringWithoutDotStartPrefixSuffix.startsWith(DOT_STAR)) {
```

Comment:
Yes, that's why this is only used in `regexp_like` function (which calls `Pattern.find()`). It gives great speed improvement in DFA since it can quickly skip (fast, tight loop) through uninteresting starting characters. Generally, this corrects user suboptimal pattern when he uses `regexp_like` (see existing regex benchmarks in Presto).